src/consensus/signature_accumulator.rs

```diff
@@ -0,0 +1,343 @@
+                        public_key,
+                        signature,
+                    },
+                )
+            })
+    }
+
+    fn remove_expired(&mut self) {
+        let expiration = self.expiration;
+        self.map.retain(|_, state| {
+            if state.modified().elapsed() < expiration {
+                true
+            } else {
+                if let State::Accumulating { payload, .. } = state {
+                    error!("Expired signature accumulation of {:?}", payload)
+                }
+
+                false
+            }
+        })
+    }
+}
+
+impl<T> Default for SignatureAccumulator<T>
+where
+    T: Debug + Serialize,
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Error returned from SignatureAccumulator::add.
+#[derive(Debug, Error)]
+pub enum AccumulationError {
+    /// There are not enough signature shares yet, more need to be added. This is not a failure.
+    #[error(display = "not enough signature shares")]
+    NotEnoughShares,
+    /// Enough signature share were already collected before and adding more has no effect. This is
+    /// not a failure.
+    #[error(display = "signature already accumulated")]
+    AlreadyAccumulated,
+    /// The signature share being added is invalid. Such share is rejected but the already collected
+    /// shares are kept intact. If enough new valid shares are collected afterwards, the
+    /// accumulation might still succeed.
+    #[error(display = "signature share is invalid")]
+    InvalidShare,
+    /// The payload failed to be serialised and can't be inserted into the accumulator. This doesn't
+    /// affect other entries already present and the accumulator can still be used normally
+    /// afterwards.
+    #[error(display = "failed to serialise payload: {}", _0)]
+    Serialise(#[error(source)] bincode::Error),
+    /// The signature combination failed even though there are enough valid signature shares. This
+    /// should probably never happen.
+    #[error(display = "failed to combine signature shares: {}", _0)]
+    Combine(#[error(from)] bls::error::Error),
+}
+
+enum State<T> {
+    Accumulating {
+        payload: T,
+        shares: HashMap<usize, bls::SignatureShare>,
+        modified: Instant,
+    },
+    Accumulated {
+        modified: Instant,
+    },
+}
+
+impl<T> State<T> {
+    fn new(payload: T) -> Self {
+        Self::Accumulating {
+            payload,
+            shares: Default::default(),
+            modified: Instant::now(),
+        }
+    }
+
+    fn add(&mut self, proof_share: ProofShare) -> Result<(T, bls::Signature), AccumulationError> {
+        match self {
+            Self::Accumulating {
+                shares, modified, ..
+            } => {
+                if shares
+                    .insert(proof_share.index, proof_share.signature_share)
+                    .is_none()
+                {
+                    *modified = Instant::now();
+                }
```

Comment:
could return early for the else branch, with NotEnough error or a new DuplicatedShare error ?